src/magic_folder/status.py

```diff
@@ -0,0 +1,180 @@
+
+import json
+import attr
+
+from zope.interface import (
+    Interface,
+    implementer,
+)
+
+from autobahn.twisted.websocket import (
+    WebSocketServerFactory,
+    WebSocketServerProtocol,
+)
+
+from twisted.application import (
+    service,
+)
+
+
+class StatusProtocol(WebSocketServerProtocol):
+    """
+    Speaks the server side of the WebSocket status protocol, usually
+    mounted at /v1/status from our web API.
+
+    This is authenticated with the same Bearer token as the rest of
+    the /v1 API.
+    """
+
+    def onOpen(self):
+        """
+        WebSocket API: successful handshake
+        """
+        self.factory._status.client_connected(self)
+
+    def onClose(self, wasClean, code, reason):
+        """
+        WebSocket API: we've lost our connection for some reason
+        """
+        self.factory._status.client_disconnected(self)
+
+    def onMessage(self, payload, isBinary):
+        """
+        WebSocket API: a message has been received from the client. This
+        should never happen in our protocol.
+        """
+        pass
+
+
+class StatusFactory(WebSocketServerFactory):
+    """
+    Instantiates server-side StatusProtocol instances when clients
+    connect.
+    """
+    protocol = StatusProtocol
+
+    def __init__(self, status):
+        """
+        :param WebSocketStatusService status: actual provider of our
+            status information. The protocol will use this to track
+            clients as they connect and disconnect.
+        """
+        self._status = status
+        WebSocketServerFactory.__init__(self, server="magic-folder")
+
+
+class IStatus(Interface):
+    """
+    An internal API for services to report realtime status
+    information. These don't necessarily correspond 1:1 to outgoing
+    messages from the status API.
+    """
+
+    def upload_started():
+        """
+        One or more items are now in our upload queue.
+        """
+
+    def upload_stopped():
+        """
+        No items are in the upload queue.
+        """
+
+
+@attr.s
+@implementer(service.IService)
+@implementer(IStatus)
+class WebSocketStatusService(service.Service):
```

Comment:
This doesn't do anything service-like. I guess it is a service so we can add to a service hierarchy?

One thing we may want to consider (though it may be unnecessary if they are being closed by other parts of the hierarchy) is closing all open connections when the service is stopped.

Reply:
Yeah, I agree it doesn't necessarily NEED to be a service for this one thing .. but conceptually it definitely seems service-like to me. And yeah cleanly closing connections on shutdown sounds like a good use-case too.